File: lib/haskell/Digit.hs
module Digit where

digits :: Integer -> [Integer]
digits x | x < 10 = [x]
digits x = (x `rem` 10):(digits $ x `div` 10)

nrDigits :: Integer -> Integer
nrDigits x | x < 10 = 1
nrDigits x = 1 + (nrDigits $ x `div` 10)

fromDigits :: [Integer] -> Integer
fromDigits = go 1 . reverse
  where
    go :: Integer -> [Integer] -> Integer
    go _ [] =  0
    go b (d:ds) = b * d + go (10*b) ds

File: lib/c++/digit.cpp
#include <stdbool.h>

#include "digit.h"

/*
 * Gets the number of digits in a given number.
 */
int nr_of_digits(int p) {
  if (p < 10) { return 1; }
  return nr_of_digits(p / 10) + 1;
}

/*
 * Calculates the number represented by the first n digits in an array of digits.
 * Whether the numbers in the array actually represent digits is not checked.
 */
int from_digits(int digit[], int n) {
  int j; int num = 0;
  for (j = 0; j < n; j++){
    num += digit[j] * pow(10, n - 1 - j);
  }
  return num;
}
